File: Kalorie/Kalorie/Features/Export/ExportViewModel.swift
//
//  ExportViewModel.swift
//  Kalorie
//
//  Created by Josef Antoni on 19.09.2026.
//

import Foundation

struct ExportedFile: Identifiable {
    let url: URL
    var id: URL { url }
}

final class ExportViewModel: ObservableObject {

    enum State {
        case idle
        case generating
    }

    // MARK: - Properties

    @Published private(set) var state: State = .idle
    @Published var fromDate: Date
    @Published var toDate: Date
    @Published var format: FoodExportFormat = .pdf
    @Published var exportedFile: ExportedFile?
    @Published var alertItem: AlertItem?

    private let mealTypes: [MealTypeDomain]
    private let generateFoodExport: any GenerateFoodExportUseCaseProtocol

    var isExportDisabled: Bool {
        state == .generating || Calendar.current.startOfDay(for: fromDate) > Calendar.current.startOfDay(for: toDate)
    }

    // MARK: - Init

    init(
        mealTypes: [MealTypeDomain],
        generateFoodExport: any GenerateFoodExportUseCaseProtocol,
        now: Date = .now
    ) {
        let calendar = Calendar.current
        self.mealTypes = mealTypes
        self.generateFoodExport = generateFoodExport
        self.fromDate = calendar.date(from: calendar.dateComponents([.year, .month], from: now)) ?? now
        self.toDate = now
    }

    // MARK: - Functions

    @MainActor
    func onExportTapped() async {
        guard !isExportDisabled else { return }
        state = .generating
        defer { state = .idle }
        do {
            let url = try await generateFoodExport(from: fromDate, to: toDate, format: format, mealTypes: mealTypes)
            exportedFile = ExportedFile(url: url)
        } catch {
            Log.error(error, category: Constants.LogCategory.export)
            alertItem = AlertItem(title: L10n.Common.errorUnknown)
        }
    }

    @MainActor
    func onShareFinished() {
        if let url = exportedFile?.url {
            try? FileManager.default.removeItem(at: url)
        }
        exportedFile = nil
    }
}

File: Android/app/src/main/java/antoni/kalorie/features/export/ExportViewModel.kt
package antoni.kalorie.features.export

import androidx.lifecycle.ViewModel
import antoni.kalorie.R
import antoni.kalorie.core.models.FoodExportFormat
import antoni.kalorie.core.models.MealTypeDomain
import antoni.kalorie.core.usecases.GenerateFoodExportUseCaseProtocol
import antoni.kalorie.core.utils.AlertItem
import antoni.kalorie.core.utils.Constants
import antoni.kalorie.core.utils.Log
import java.io.File
import java.time.Instant
import java.time.ZoneId
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow

data class ExportedFile(val file: File, val format: FoodExportFormat)

class ExportViewModel(
    private val mealTypes: List<MealTypeDomain>,
    private val generateFoodExport: GenerateFoodExportUseCaseProtocol,
    now: Instant = Instant.now(),
    private val zone: ZoneId = ZoneId.systemDefault(),
) : ViewModel() {

    // MARK: - State

    enum class State {
        IDLE,
        GENERATING,
    }

    // MARK: - Properties

    private val _state = MutableStateFlow(State.IDLE)
    val state: StateFlow<State> = _state
    val fromDate = MutableStateFlow(now.atZone(zone).toLocalDate().withDayOfMonth(1).atStartOfDay(zone).toInstant())
    val toDate = MutableStateFlow(now)
    val format = MutableStateFlow(FoodExportFormat.PDF)
    val exportedFile = MutableStateFlow<ExportedFile?>(null)
    val alertItem = MutableStateFlow<AlertItem?>(null)

    val isExportDisabled: Boolean
        get() = _state.value == State.GENERATING ||
            fromDate.value.atZone(zone).toLocalDate().isAfter(toDate.value.atZone(zone).toLocalDate())

    // MARK: - Functions

    suspend fun onExportTapped() {
        if (isExportDisabled) return
        _state.value = State.GENERATING
        try {
            val file = generateFoodExport(fromDate.value, toDate.value, format.value, mealTypes)
            exportedFile.value = ExportedFile(file, format.value)
        } catch (error: CancellationException) {
            throw error
        } catch (error: Exception) {
            Log.error(error, Constants.LogCategory.EXPORT)
            alertItem.value = AlertItem(titleRes = R.string.common_error_unknown)
        } finally {
            _state.value = State.IDLE
        }
    }

    fun onShareFinished() {
        exportedFile.value = null
    }
}
